subprojects/core/src/main/java/org/gradle/util/internal/Java10ClassReader.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.util.internal;
+
+import org.gradle.api.JavaVersion;
+import org.objectweb.asm.ClassReader;
+
+/**
+ * A <b>temporary workaround</b> for reading Java 10 class files. This special class reader will pretend that Java 10
+ * classes are Java 9 classes, making it possible to read them.
+ */
+public class Java10ClassReader extends ClassReader {
```

Comment:
I'd probably name this "PatchedClassReader" and keep it around, as we will probably need this again and again. E.g. https://github.com/gradle/gradle/issues/3935 will need to add some code here as well.